index.bs

```diff
@@ -4421,7 +4441,8 @@ the {{WritableStream}}'s public API.
  1. [=Reject=] |stream|.[=WritableStream/[[inFlightCloseRequest]]=] with |error|.
  1. Set |stream|.[=WritableStream/[[inFlightCloseRequest]]=] to undefined.
  1. Assert: |stream|.[=WritableStream/[[state]]=] is "`writable`" or "`erroring`".
- 1. If |stream|.[=WritableStream/[[pendingAbortRequest]]=] is not undefined,
+ 1. If |stream|.[=WritableStream/[[pendingAbortRequest]]=] is not undefined and |error| is not an
```

Comment:
This kind of switching on error type is a bit unusual. What is the intention here? What kind of test would pass with this change that would fail without it?

Reply:
```
let reject;
close_promise = new Promise((res, rej) => reject = rej);
const ws = new WritableStream({
  start(c) {
    c.signal.addEventListener('abort', () => {
      const abort_error = new DOMException('foobar', 'AbortError');
      reject(abort_error);
    });
  }, close() {
    return close_promise;
  }
});

const writer = ws.getWriter();
writer.close();
const reason = Exception('hoge');
writer.abort(reason);
```

In this example, I think it's natural that `writer.closed` will be rejected by `reason`, but without this clause it'll be rejected with `abort_error`.

Another solution would be to expose `pendingAbortRequest.reason` in the controller so that the underlying sink can reject the result of `close()` with that.

Reply:
> In this example, I think it's natural that writer.closed will be rejected by reason, but without this clause it'll be rejected with abort_error.

Hmm, I'm not sure I agree. The close() underlying sink hook explicitly said it failed to close (i.e. it returned a promise rejected with an "AbortError" DOMException, indicating that it tried to close but closing got aborted). I think that failure should be propagated to the `closed` promise...

Reply:
But in this case, the operation fails because of the abort call coming from the producer, so I think it should be rejected with the abort reason. With the other solution, that code would look like:

```
let reject;
close_promise = new Promise((res, rej) => reject = rej);
const ws = new WritableStream({
  start(c) {
    c.signal.addEventListener('abort', () => {
      reject(c.abortReason);
    });
  }, close() {
    return close_promise;
  }
});

const writer = ws.getWriter();
writer.close();
const reason = Exception('hoge');
writer.abort(reason);
```

Reply:
I think I prefer `reject(c.abortReason)`.

Reply:
> But in this case, the operation fails because of the abort call coming from the producer, so I think it should be rejected with the abort reason

Well, but it only "failed" because the underlying sink said it failed. The underlying sink should just not return a rejection from the `close()` method at all, IMO. Then the usual logic [where calling `abort(r)` during close causes `closed` to reject with `r`](https://github.com/web-platform-tests/wpt/blob/master/streams/writable-streams/aborting.any.js#L231-L245) will take over.

Reply:
Hmm. So you'd do `resolve()` instead? Or something like `return close_promise.catch(() => {})`? 🤔

Reply:
I think I need a non-artificial example. Why did you reject the closed promise in the first place? I would think that the signal would have no impact on the promise returned from close, but instead would be used while opening a file or writing to it.

E.g.

```js
let fd;
const ws = new WritableStream({
  start(c) {
    fd = await fs.open(filePath, { signal: c.signal });
  },
  write(chunk, c) {
    await fd.write(chunk, { signal: c.signal });
  }
  close() {
    await fd.close(); // no signal; closing isn't abortable
  }
});
```

If you changed the line to `await fd.close({ signal: c.signal })`, i.e. you let someone abort in the middle of closing---then yes, I'd expect `ws.closed` to reject with an "AbortError" DOMException, because we failed to close the stream due to the closing process being aborted!

Reply:
Let me quote @vasilvv's example for WebTransport.

>  Imagine I have a web app where I can do drawing online.  Every 100ms or so, I send the entirety of the drawing on a dedicated WebTransport stream.  Note that if a network connection becomes spotty for a short period of time, this will cause a lot of streams to be open at once, so I want to keep only the three most recent streams open, and reset all streams that are older than the last three.

Here we want to map [the reset operation](https://w3c.github.io/webtransport/#stream-reset) to `abort()`. We are planning to resolve the promise returned by `close()` when the the last chunk is ack-ed, and until then the close operation is always cancellable - by just sending a RESET_STREAM frame.

Reply:
OK, so in that case, it seems like you can avoid the problem by not rejecting the promise returned by close when the reset operation is done. Returning a rejected promise from close indicates the closing failed, which it doesn't sound like is accurate in that case.

Reply:
Reverted the edit. I'm not sure if we don't need to access `abortReason`.